### PR TITLE
bugfix: ensure that at::{dispatch_key}:: API gets external linkage

### PR DESCRIPTION
**Pull request by bdhirsh**
This should allow external C++ files that aren't compiled into `libtorch.so`/`libtorch_cpu.so` (including all of fbcode) to use fast path functions like `at::cpu::add()`, which skip the dispatcher.

So, after spending way too much time trying to figure out why I was getting linker errors when calling `at::meta::{op}` and `at::cpu::{op}` from C++ test files, I realized that we're not including the header files for C++ for the namespaced operator definitions. I.e. `RegisterCPU.cpp`, which provides definitions for the `at::cpu::{op}` fast path functions, wasn't including the `CPUFunctions.h` header.

Why that breaks stuff: the `CPUFunctions.h` header file is what marks each function with the `TORCH_API` macro, so without including it, when we build `libtorch.so` and `libtorch_cpu.so`, the compiler will look at the definition in `RegisterCPU.cpp`, not see a `TORCH_API`, and decide that the function should get internal linkage.

An alternative would be to directly mark the function definitions in `RegisterCPU.cpp` with `TORCH_API`, but this seemed cleaner.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59839 move all external kernels into a class for better compiler error messages
* #58065 add a boxed CPU fallback kernel
* #59115 beef up at::_ops API
* #58570 generate C++ API for meta functions using at::meta::
* **#58569 bugfix: ensure that at::{dispatch_key}:: API gets external linkage**

Differential Revision: [D28711300](https://our.internmc.facebook.com/intern/diff/D28711300)

### CODE REVIEW

File: tools/codegen/gen_backend_stubs.py [hunk 1]
```diff
@@ -205,6 +205,7 @@ def make_file_manager(install_dir: str) -> FileManager:
                 'extra_cuda_headers': '',
                 'legacy_th_headers': '',
                 'external_backend_headers': external_backend_headers,
+                'namespaced_headers': '',
```
Comment (bdhirsh) [inline tools/codegen/gen_backend_stubs.py:208]: I haven't actually started codegen'ing the `XLAFunctions.h` file that exposes the `at::xla::{op}` fast path API - when we eventually do that we'll need the same logic here.

**Review by ezyang — APPROVED** [review]
Nice catch, sorry about sending you on a wild bugfix hunt.

**Comment by bdhirsh** [comment]
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28711300).

**Comment by ezyang** [comment]
need this for one of my PRs

**Comment by bdhirsh** [comment]
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28711300).

**Comment by bdhirsh** [comment]
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28711300).

**Comment by bdhirsh** [comment]
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28711300).

**Comment by bdhirsh** [comment]
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28711300).

**Comment by bdhirsh** [comment]
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28711300).

**Comment by bdhirsh** [comment]
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28711300).

**Comment by bdhirsh** [comment]
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28711300).

**Comment by bdhirsh** [comment]
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28711300).

**Pull request closed without merging**